pipeline.py

```diff
@@ -1,43 +1,90 @@
 from prefect import task, Flow
 
 import pandas as pd
+import os
 import fsspec
-
-
-url_pattern = 'https://www.ncei.noaa.gov/data/sea-surface-temperature-optimum-interpolation/access/avhrr-only/{{ %Y%m }}/avhrr-only-v2.{{ %Y%m%d }}.nc'
-
-date_range = {'start': '1981-09-01', 'end': '2019-08-11', 'freq': 'D'}
-dates = pd.date_range(**date_range)
-files = dates.strftime(url_pattern)
-
-
+import zarr
+import itertools
+
+# https://alexwlchan.net/2018/12/iterating-in-fixed-size-chunks/
+def chunked_iterable(iterable, size):
+    it = iter(iterable)
+    while True:
+        chunk = tuple(itertools.islice(it, size))
+        if not chunk:
+            break
+        yield chunk
+
+def src_url_from_key(key):
+    source_url_pattern = 'https://www.ncei.noaa.gov/data/sea-surface-temperature-optimum-interpolation/v2.1/access/avhrr/{yyyymm}/oisst-avhrr-v02r01.{yyyymmdd}.nc'
+    return source_url_pattern.format(yyyymm=key.strftime('%Y%m'), yyyymmdd=key.strftime('%Y%m%d'))
+    
+cache_location = 'gs://pangeo-scratch/rabernat/pangeo_smithy/oisst-avhrr-v02r01-cache/'
+    
+    
 @task
-def download(source):
-    import wget
-
-    wget.download(source)
-
-    return target
+def download(key):
+    source_url = src_url_from_key(key)
+    os.path.join(cache_location, str(hash(key)))
+    target_url = os.path.join(cache_location, str(hash(key)))
+    
+    # there is probably a better way to do caching!
+    try:
+        fsspec.open(target_url).open()
+        return target_url
+    except FileNotFoundError:
+        pass
+    
+    with fsspec.open(source_url, mode='rb') as source:
+        with fsspec.open(target_url, mode='wb') as target:
+            target.write(source.read())
+    return target_url
 
 
 @task
-def combine(sources):
+def combine_and_write(sources, target, append_dim, first=True):
     import xarray as xr
-
-    ds = xr.open_mfdataset(sources).chunk({'time': 5})
+    dsets = []
+    # while debugging this, I had itermittent fsspec / hdf5 read errors related to
+    # "trying to read from a closed file"
+    # but they seem to have gone away for now
+    double_open_files = [fsspec.open(url).open() for url in sources]
+    ds = xr.open_mfdataset(double_open_files, combine='nested', concat_dim=concat_dim)
```

Comment:
@martindurant - would love some feedback on whether this is the "right" way to use fsspec here.

Reply:
Sure, this should work. Note that the files stay open until explicitly closed or garbage collected at the end of the function. The code to make `.open()` work is not the prettiest I've ever written. Also note that `fsspec.open_files` is maybe more efficient for a list of URLs than `open` on each.  
fsspec should maybe have a "combined context" that works on more than one file, so you can do

```
with fsspec.open(sources) as double_open_file:
    ...
```

Reply:
> . Also note that `fsspec.open_files` is maybe more efficient for a list of URLs than `open` on each.

But then wouldn't I have to just loop through the OpenFile objects and call `.open()` again on each one?



> fsspec should maybe have a "combined context" that works on more than one file, so you can do

Yes this would be good. I couldn't figure out a way to keep multiple files open with the context manager without resorting to some horrible nested `with` structure.

Reply:
> But then wouldn't I have to just loop through the OpenFile objects

Yes, but you would skip the URL inference stuff. Maybe it makes little difference.

Reply:
https://github.com/intake/filesystem_spec/issues/357